### ISSUE TITLE
Frontend development discussion - input needed

### ISSUE THREAD

**Issue by tommueller** [vitotai/BrewPiLess#21]
I started working on a facelift for BrewPiLess' frontend. @vitotai added the files to the repository, see for this #18. The latest code can be found in my fork of the repository: https://github.com/tommueller/BrewPiLess

Because I don't have a lot of experience in being a BrewPiLess user I do need some input from users concerning the UI elements. Right now I am especially wondering about: 
- which are the most important information?
  - the chart
  - the gravity pane in the top
  - the menu in the bottom
- should the two parts of the page (plot and the settings on the bottom) stay on one page, or can they be split?
- any other ideas, suggestions?

I already cleaned up the code a little bit and started to make a clean, simple, material design layout for the upper part (plot) of the page, which looks like this right now:
![auswahl_001](https://user-images.githubusercontent.com/2970691/29244940-43a35c5e-7fc8-11e7-8316-1bd8210517b6.png)

So please, any input is welcome!

**Comment by vitotai** [comment]
That looks cool. Really.

In addition to mimicking the original BrewPi, another reason to put temperature control part under the chart is that it informative when beer profile is used. The beer profile chart and table provide the information of temperature schedule. (It's not pretty. I know.)

The LCD is refreshed every 5 seconds while the chart is updated every minute. The LCD also provides some other information about the system, like temperature control mode and idle/cooling/heating time.   A small piece of Javascript can used to retrieve the data, so that the information can be shown in other form.

The "iSpindel name" is not necessary, IMO. It's there because it's available.

**Comment by zandoga** [comment]
Looks very good! I could not figure out what "state" is used for.

**Comment by vitotai** [comment]
When the mouse cursor hovers over the line of chart, "state" will show the state of BrewPi, like "Idle", "Cooling", "Heating", "Prepare to Cool", and etc.

**Comment by tommueller** [comment]
Thanks for the kind words.

The LCD was not meant to be chucked, I simply did not find a good spot to put it so far :)

> When the mouse cursor hovers over the line of chart, "state" will show the state of BrewPi, like "Idle", "Cooling", "Heating", "Prepare to Cool", and etc.

what exactly do you hover and where does it show?

**Comment by tommueller** [comment]
Any chance that you have a local version of the server so that I can test the js as well?

**Comment by vitotai** [comment]
> what exactly do you hover and where does it show?
When the mouse cursor hovers over the lines in the chart, the "state" will be REPLACED but the state string.

> Any chance that you have a local version of the server so that I can test the js as well?
I did use some simple PHP scripts to test the JS, but they are not maintained.
The original BrewPi core uses JSON format with an additional character at the beginning of output. bwf.js is used to be the interface between Javascript and BrewPi core. The purpose of testcmd.htm is to test the response of BrewPi core. (Later, I found it is useful to set additional setting and put it in the release code.)
I tried to shift the loading to the brewer, so most of the data is in JSON format and ESP8266 just pass what it is to JS mot of the time. 

Chart data is an exception for reduction of network traffic and processing of data by ESP8266. It is still simple in the JS part. Here is my PHP script to pass a saved log to JS.( It was quite dirty and complicated when I was developing the code for periodical chart updating. However, I don't even keep the code after finishing.)

`<?php
header('Content-Type: application/octet-stream');
header('Cache-Control: no-cache'); // recommended

$pmtime=0;
$linenum=3;
if(array_key_exists("offset",$_GET)){
	$offset=$_GET["offset"];
}else{
	$offset=0;
}

$filename ="coldsnap0509";

if(filesize($filename) > $offset){
$bin=fopen($filename,"r");
if($offset > 0){
	fseek($bin,$offset);
	echo fread($bin,filesize($filename)-$offset);
}else{
	echo fread($bin,filesize($filename));
}
}
?>`

**Comment by tommueller** [comment]
Ok thanks!
One more question, does the chart have to stay at a fixed width of 800px, or can the script handle dynamic sizes?
Cheers

**Comment by allthepies** [comment]
The recording div is populated when logging is enabled.  The logging filename is shown.

**Comment by tommueller** [comment]
Ok, this got some progress today. You can test the current version if you download the HTML from my fork. It's still very much work in progress, I am just doing the chart in the first run. I figured out that it can have any size and __all the question asked so far, are now clear__.

I do have one more question though. What is shown as `fgstate`?

**Comment by vitotai** [comment]
That is a small LED indicating the stability of gravity reading.
     0: "red",
     12: "orange",
      24: "yellow",
      48: "green"

The numbers denote the number of hours during which the gravity changes only a little, default to 1.

**Comment by vitotai** [comment]
I just add a indicator for WiFi signal.

![wifisignal](https://user-images.githubusercontent.com/15953213/29498360-5c95527a-85af-11e7-801a-d8198fe33573.jpg)

**Comment by tommueller** [comment]
Cool, I will have some time over the upcoming weekend / beginning of next week and hope to be able to have a first running version ready then.

**Comment by tommueller** [comment]
Some progress, will need a tester soon who has a version running with more then one temp sensor like me :) anyone?

**Comment by allthepies** [comment]
I'll give it a go, will try on my testbed and then on my fridge setup which has three sensors, beer, fridge and room.

Do I just build/upload your fork using platformio or do I need to do some magic first ?

**Comment by tommueller** [comment]
@allthepies great thanks. Gimme 2 more days, I'll let you know when it's done. You just need download the html from the dist folder and upload it using the filemanager.

**Comment by allthepies** [comment]
Wilko.

**Comment by vitotai** [comment]
I will test it, too. 
BTW, some people seems to like use their own names at the place of BrewPiLess. Therefore, I display the "hostname" there. It seems better for people who have more than one setup like me.

![bpl2](https://user-images.githubusercontent.com/15953213/29796556-9c7b0170-8c06-11e7-902b-1a1ea90dfa31.jpg)

**Comment by tommueller** [comment]
Ok, I pushed a first running version into my fork. I split up the main-page into 2 (chart and control), which creates some problems (dependencies between control part and chart). In order to have all the functions you need to upload both html-files from the dist folder.

Some notes:
- the lcd-text format seems to be set in the ESP8266 already? I was not able to hack into it, to format the text nicer for the new page-style
- hovering the charts shows some kind of popup? I haven't found the time to track the corresponding css, so right now hovering the chart makes the page a bit jumpy (because the "popup" is just appended to the page, scroll down to see the content)
- the control-page does not seem to work, because of some interdepencies with the chart. is that necessary? or is it okay to split the parts?

here is a screenshot for the ones not wnating to test it:
![bildschirmfoto vom 2017-09-01 14-13-14](https://user-images.githubusercontent.com/2970691/29969318-d173edd0-8f1f-11e7-8e2e-59de8cf79ede.png)

Looking forward to some feedback and maybe some help on fixing the open issues.

**Comment by allthepies** [comment]
Looks good Tom, I'll give it a go over the w/e and report back.

Just so I understand, installation involves uploading the two html files from your repo htmljs/dist to a BPL instance using the BPL file manager ?

**Comment by tommueller** [comment]
Cheers!

> Just so I understand, installation involves uploading the two html files from your repo htmljs/dist to a BPL instance using the BPL file manager ?

Exactly!

**Comment by allthepies** [comment]
OK, got it up and running, looks great!

Initial comments below.

- Chart controls - toggle colours don't match the chart line colours, confusing.
Control page - Javascript error, tabs don't work
Uncaught TypeError: Cannot set property 'onmouseover' of null
    at init (control.html:1)
    at onload (control.html:1)
	
Control page - version different to that shown on main page  (main:  v2.3.2, control: v2.1)
Control page:  Shows "BrewPiless" banner rather than the hostname as per the main page.
Control page:  Colour mismatch, banner green, control pane is the original blue, could do with harmonising.

Navigation - have to explicitly open each sub-page in new tab otherwise main page is switched.  UI debate and I know magic new tabs is frowned upon, one for discussion.
Main page:  No wifi strength shown, just "wifi: "
Main page: When logging enabled, log filename is shown under the wifi strength but just the name.  Prefix with "Log filename:" or similar.

**Comment by vitotai** [comment]
The LCD information can be extracted, and display in other way. A few regular expression would get the needed information from the plain text.

How about we break down the information into these fields:
BeerTemp
BeerSet
FridgeTemp
FridgeSet
RoomTemp
ControlMode
ControlState 
ControlStateSince ( the "time" parts)

Given these words are common, I would like to use a prefix, "lcd". How do you think?

**Comment by tommueller** [comment]
@allthepies thanks for testing and the feedback, some comments below:

> Chart controls - toggle colours don't match the chart line colours, confusing.

I will look into this.

> Control page - Javascript error, tabs don't work
Uncaught TypeError: Cannot set property 'onmouseover' of null
at init (control.html:1)
at onload (control.html:1)

Control page - version different to that shown on main page (main: v2.3.2, control: v2.1)
Control page: Shows "BrewPiless" banner rather than the hostname as per the main page.
Control page: Colour mismatch, banner green, control pane is the original blue, could do with harmonising.

The control page is not working because it has some dependencies on the chart. Since I never used it I am not sure if this is a useful feature or if I can remove the code causing it. I need some input from you guys for this.

>  Navigation - have to explicitly open each sub-page in new tab otherwise main page is switched. UI debate and I know magic new tabs is frowned upon, one for discussion.

My plan was to integrate the other pages into the same header, footer style and have page changes, because I don't like opening tabs, but this can/should be debated here!

> Main page: No wifi strength shown, just "wifi: "

did you upgrade to version 2.3.2? You need to have the latest version to support this.

> Main page: When logging enabled, log filename is shown under the wifi strength but just the name. Prefix with "Log filename:" or similar. 

> Main page: Gravity axis value issue, see image below, the top gravity reading shows 0.54 which is less that than the value below it.

I don't think this a relayout issue, but rather an issue in the graph library, @vitotai ? I did not change anything on the graph itself.

**Comment by tommueller** [comment]
> How about we break down the information into these fields:
BeerTemp
BeerSet
FridgeTemp
FridgeSet
RoomTemp
ControlMode
ControlState
ControlStateSince ( the "time" parts)

> Given these words are common, I would like to use a prefix, "lcd". How do you think?

@vitotai yes, sounds perfect! I am not 100% sure if the regex stuff should be done within the js-code or if it is in the BPL code, need some directions on what to do here, because I did not find where the lcd-text is generated.

**Comment by allthepies** [comment]
I tested with the latest version i.e. 2,3.2  ( I always adopt new versions quickly after testbed proving :) )

re: control page, that is the key page for BPL.  Without it we cannot change the temp settings  (well I can with my Alexa skill but that's a mute point ;) )

**Comment by vitotai** [comment]
The firmware part keeps a copy of LCD display. The javascript periodically issue  "l" command to get it. There are other commands like "s" and "t" that can be used to retrieve mode and current temperatures. However, the information of last line is only available on the LCD, and the response of LCD. I prefer parsing the LCD text to extract the information since it's simple enough, and using one command is better than using two. 

One issue remains. The last line of LCD is the status of the system, it can be the following patterns.
Idle: Idling for  %dh%02dm%02d
Cooling: Cooling for %dh%02dm%02d
Heating: Heating for %dh%02dm%02d
Waiting_to_cool: Wait to Cool %dh%02dm%02d
Waiting_to_cool:Wait to heat %dh%02dm%02d
cooling_min_time: Cool Time Left %dh%02dm%02d
heating_min_time: Heat Time left %dh%02dm%02d
Waiting_for_peak_detect: Waiting for peak
Door Open
Temp. control OFF
Unknown status!


Three options I have thought:
1. keep the whole sentence. Use the whole line as one.
2. Extract the time and state, for example
Idling for  %dh%02dm%02d =>  ControlState = "Idling" , ControlStateSince = "%dh%02dm%02d"
3.Using the state index. e.g., idle is 0, cooling is 4. Defined as variable "STATES".
Idling for  %dh%02dm%02d =>  ControlState = 0, ControlStateSince = "%dh%02dm%02d"

Your opinion?

**Comment by vitotai** [comment]
I add the parsing function and modify the HTML to demonstrate.
[index.htm.zip](https://github.com/vitotai/BrewPiLess/files/1271724/index.htm.zip)

**Comment by tommueller** [comment]
@vitotai and others: before I can continue working on this (and especially the control-page) we need to decide if we
a) need to have graph and control part on one page? and
b) want a "normal" navigation or should logs/config etc. pages open in a new tab as it is right now?

My personal opinions are:
for a) split control and graph pages, because it is different functionalities
for b) I prefer to stay in a tab and navigate between pages, this is really just what I expect from a page in 2017, especially because I do not need to have the chart open when I change something in the setup

What are your opinions?

**Comment by zandoga** [comment]
You suggestion sounds good to me.

**Comment by tommueller** [comment]
@vitotai the parsing works like a charm! already included it into my fork. thanks!

I updated my repo with a new version, I think all the issues on the graph-page should be resolved now, but please test again everyone!

@all please give more opinions on [my question before](https://github.com/vitotai/BrewPiLess/issues/21#issuecomment-326736597)

**Comment by allthepies** [comment]
Hi Tom,

Just uploaded the files from your latest commit and I can see no difference.  What was supposed to have changed ?

Cleared browser several times etc.

**Comment by tommueller** [comment]
@allthepies sounds like a cache problem. did you reload with shift pressed? maybe just try again (just pushed a new version again, with a little fix).

Changes are:
- better sorting and display of information in the sidepanel, including log-filename
- fixed wifi-colors, they were confusing
- toggle colors should match the line color in graph now
- some minor style improvements

**Comment by allthepies** [comment]
I don't know what's going on Tom, it now refuses to load your files from the filesystem and is presenting the old pages from the program memory.  I've wiped the ESP8266 SPIFFS filesystem and started again but the same result.

**Comment by allthepies** [comment]
OK!  

Your index.html file has to be uploaded as index.htm  (no trailing "l") as that's what BPL looks for in the SPIFFS filesystem.   Your control.html file has to be uploaded as-is  i.e. control.html   as that's what the main page HTML links to!

Confusing eh.  Perhaps rename index.html -> index.htm and change control.html page and reference links to control.htm ?

**Comment by tommueller** [comment]
That is strange, I've always uploaded it as .html and it worked just fine. It doesn't feel like .htm is a valid file extension to me, so I changed it to html, but maybe I'll change it back ;)

Good that is working now! Thanks for testing and reporting back that issue!

@vitotai is there any reason for the .htm-file-ending? why is it working for me with .htm__l__ but not for @allthepies ? any ideas?

**Comment by allthepies** [comment]
Lots of JS errors referencing LCD text

![bplerrors](https://user-images.githubusercontent.com/1494526/29996762-5a39cbe4-8ffc-11e7-993d-fbe4b80c10d9.JPG)

**Comment by tommueller** [comment]
Mmh. That's weird. Can you try the attached file, I added some output to the console to see where the error comes from, cannot recreate it here. You have to rename it to index.htm of course.
[index.txt](https://github.com/vitotai/BrewPiLess/files/1272280/index.txt)

**Comment by allthepies** [comment]
![bplerror](https://user-images.githubusercontent.com/1494526/29996904-29ce2e34-8fff-11e7-9a4f-0f0b1a360dd7.JPG)

**Comment by vitotai** [comment]
1. The index file should be "index.htm", I am not sure if "index.html" works or not. I've played with it for a while, and decide to use index.htm. Just forget why. Or maybe there isn't any reason.

2. ESP8266 or the upload program seems to have trouble handling BIG file. I was stuck at the issue for a while when I used AsyncWebServer. The default synchronized web server seems to work better, but it still has this issue. The symptom is the corrupted file. **You need to "reload" again. Or upload the file again.** It might be caused by upload or download.
Files smaller than 80k are safer. I never succeed on files over 130k.

**Comment by tommueller** [comment]
@vitotai great thanks for the input. I renamed files in the output folder to .htm! Let me know when compression is an option, then I'll update the grunt-task to autocompress the files.

Do you have any idea why the parsing of the lcd-text is failing for @allthepies ? It works like a charm for me ... Don't know what's failing for him, but I only have one temperature (room), maybe it's an issue with other values?

**Comment by allthepies** [comment]
My testbed has no temperature probes attached, perhaps that's the gotcha ?

**Comment by vitotai** [comment]
I know the cause of JS error. I assumed the time is alway in %dh%02dm%02d, it turns out the "hour" doesn't exist. Give me a few minutes to verify the corrected one.

**Comment by vitotai** [comment]
[index.htm.gz](https://github.com/vitotai/BrewPiLess/files/1272310/index.htm.gz)

The regular expressions are updated. Some more exception protection is also added.
The gzip is supported now in the source. (not pre-built images)
I recommend using highest compression ration, -9. so
`
gzip -9 index.htm
`
and just upload the index.htm.gz and let it be exact the name 'index.htm.gz' .

**Comment by tommueller** [comment]
@vitotai sorry cannot find the changes in your file (I use seperate files for development and merge them in the end, so I cannot just override with your file). Can you maybe point me to them?
Also gzip will only be working once I updated my BPL, right? I only use your bins, so I'll have to wait for that.

**Comment by allthepies** [comment]
Yes, you need the BPL version Vito's just commited to support SPIFFS served gzip files.

I've updated my test bed and am ready whenever you are...

**Comment by vitotai** [comment]
You need new image to support GZ files. 

Without the gzip functionality, you have to use plain index.htm, not index.htm.gz.

**Comment by tommueller** [comment]
@vitotai thanks for the bin

I do get the error now myself, but I cannot find the changes in your uploaded .htm file. Can you please post the differences? Or the functions that I need to override?

Thanks!

**Comment by vitotai** [comment]
[index.htm.gz](https://github.com/vitotai/BrewPiLess/files/1272383/index.htm.gz)
I used old file. 

The updated functions here:
`

        function parseLcdText(lines) {
            var status = {};
            var modePatterns = {
                b: /Mode\s+Beer\s+Const/i,
                f: /Mode\s+Fridge\s+Const/i,
                p: /Mode\s+Beer\s+Profile/i,
                o: /Mode\s+Off/i
            };
            var modes = Object.keys(modePatterns);
            status.ControlMode = "i";
            for (var m = 0; m < modes.length; m++) {
                if (modePatterns[modes[m]].test(lines[0])) {
                    status.ControlMode = modes[m];
                    break;
                }
            }
            status.ControlState = i;
            var tempRE = /\s*(\w+)\s+(.+)\s+(.+)\s+.+([CF])\s*$/;
            for (var i = 1; i < 3; i++) {
                var temps = tempRE.exec(lines[i]);
                status[temps[1] + "Temp"] = temps[2];
                status[temps[1] + "Set"] = temps[3];
                status.format = temps[4];
            }
            var i = 0;
            var statePatterns = [
                /Idling\s+for\s+(\S+)\s*$/i,
                /control\s+OFF/i,
                /Door\s+Open/i,
                /Heating\s+for\s+(\S+)\s*$/i,
                /Cooling\s+for\s+(\S+)\s*$/i,
                /Wait\s+to\s+Cool\s+(\S+)\s*$/i,
                /Wait\s+to\s+Heat\s+(\S+)\s*$/i,
                /Wait\s+for\s+Peak/i,
                /Cool\s+Time\s+left\s+(\S+)\s*$/i,
                /Heat\s+Time\s+left\s+(\S+)\s*$/i
            ];
            status.ControlStateSince = "";
            for (i = 0; i < statePatterns.length; i++) {
                var match = statePatterns[i].exec(lines[3]);
                if (match) {
                    if (typeof match[1] !== "undefined") status.ControlStateSince = match[1];
                    break;
                }
            }
            status.ControlState = i;
            status.StatusLine = lines[3];
            return status;
        }

        function processLcdText(lines) {
            Q("#status-pane .error").style.display = "none";
            var status = parseLcdText(lines);
            var ModeString = {
                o: "OFF",
                b: "Beer Constant",
                f: "Fridge Const",
                p: "Beer Profile",
                i: "Invalid"
            };
            Object.keys(status).map(function(key, i) {
                var div = Q("#lcd" + key);
                if (div) {
                    if (key == "ControlMode") div.innerHTML = ModeString[status[key]];
                    else if (key == "ControlState") div.innerHTML = (status[key] < STATES.length) ? STATES[status[key]].text : "Unknown State";
                    else div.innerHTML = status[key];
                }
            });
        }
`

**Comment by tommueller** [comment]
Thanks @vitotai !

I fixed the code in my branch and it should be ready for testing @allthepies

**Comment by allthepies** [comment]
:+1: 

Temps now displaying OK and no lcd console error.

However, hovering over the Wifi singnal strength indicator gives a JS error (see image).  Should display the %age strength.

![bplwifierror](https://user-images.githubusercontent.com/1494526/29997789-5137e20c-9014-11e7-82eb-30d60988416a.JPG)

**Comment by tommueller** [comment]
Ah well spotted, the mouseover should still work though, the value is added as HTML-title attribute now, so it's popping up with a delay of about 1s!

I removed the lines that caused the error and fix is pushed. Thanks

**Comment by tommueller** [comment]
Allright, did quite some work this morning to clean up and split up the scripts and start to get the control page done. I have just pushed a new version, please test the index and also the control, some notes:

- graph-page (index.htm) should be almost done by now! I fixed some smaller issues, added ui-validation for gravity-input, styled the inputs etc. please have a go to try and test all the functionality!
- I started working on the control-page
  - it loads hostname and version from BPL now
  - added general styles, replaced the tabs with better looking ones
  - I moved around quite some scripts, to get a cleaner and smaller file, so please test funcationality!
  - @vitotai when I am on the control-page the lcd-writing apparently always sends "Controller not updating data" back, which I show in a notification on the bottom, because I treat it as an error. Why is this and should I just dismiss the notification or is there a problem with the script?

**Comment by tommueller** [comment]
Okay. Control-Page is now also done, as far as I can tell. So please: testing, testing, testing! And let me know if you don't like any of the applied styles!
Cheers!

**Comment by vitotai** [comment]
> when I am on the control-page the lcd-writing apparently always sends "Controller not updating data" back, which I show in a notification on the bottom, because I treat it as an error. Why is this and should I just dismiss the notification or is there a problem with the script?


The init() function were created by merging control and graph pages. To update the LCD, the script periodically issuing "l" command. The script also check if any response from the server. It is not necessary for pages which don't have LCD.


`
function init() {
  loaded();
  var gotMsg=true;

  BWF.init({
   /* this is not necessary
    onconnect:function(){
      BWF.send("l");
      setInterval(function(){
        if(!gotMsg) controllerError();
        BWF.send("l");
        gotMsg=false;
      },5000);
    }, */
    error:function(e){
      communicationError();
    },
    handlers:{
      V:function(c){
        if(typeof c["nn"]!="undefined"){
          Q("#hostname").innerHTML = c["nn"];
        }
        if(typeof c["ver"]!="undefined"){
          if(JSVERION != c["ver"]) alert("Version Mismatched!. Reload the page.");
          Q("#verinfo").innerHTML = "v" + c["ver"];
        }
      }
    }
  });
};
`

**Comment by tommueller** [comment]
Yes, I put this in to have hostname and version info available to put in in the header. But I see which part I can get rid of, thanks!

**Comment by vitotai** [comment]
It was over 100 Fahrenheit yesterday, and I found the regex will fail if Fahrenheit is used. Please update the regex. 

`
  var tempRE = /\s*([a-zA-Z]+)\s*([-\d\.]+)\s+([-\d\.]+)\s+(\S+[CF])\s*$/i;
`

BTW, how do think about adding the temperature unit?
`
 else if (key.match(/[Temp|Set]$/)) div.innerHTML = status[key] + status.tempUnit;
`

The full code:

`
        function parseLcdText(lines) {
            var status = {};
            var modePatterns = {
                b: /Mode\s+Beer\s+Const/i,
                f: /Mode\s+Fridge\s+Const/i,
                p: /Mode\s+Beer\s+Profile/i,
                o: /Mode\s+Off/i
            };
            var modes = Object.keys(modePatterns);
            status.ControlMode = "i";
            for (var m = 0; m < modes.length; m++) {
                if (modePatterns[modes[m]].test(lines[0])) {
                    status.ControlMode = modes[m];
                    break;
                }
            }
            status.ControlState = i;
            var tempRE = /\s*([a-zA-Z]+)\s*([-\d\.]+)\s+([-\d\.]+)\s+(\S+[CF])\s*$/i;
            for (var i = 1; i < 3; i++) {
                var temps = tempRE.exec(lines[i]);
                status[temps[1] + "Temp"] = temps[2];
                status[temps[1] + "Set"] = temps[3];
                status.tempUnit = temps[4];
            }
            var i = 0;
            var statePatterns = [
                /Idling\s+for\s+(\S+)\s*$/i,
                /control\s+OFF/i,
                /Door\s+Open/i,
                /Heating\s+for\s+(\S+)\s*$/i,
                /Cooling\s+for\s+(\S+)\s*$/i,
                /Wait\s+to\s+Cool\s+(\S+)\s*$/i,
                /Wait\s+to\s+Heat\s+(\S+)\s*$/i,
                /Wait\s+for\s+Peak/i,
                /Cool\s+Time\s+left\s+(\S+)\s*$/i,
                /Heat\s+Time\s+left\s+(\S+)\s*$/i
            ];
            status.ControlStateSince = "";
            for (i = 0; i < statePatterns.length; i++) {
                var match = statePatterns[i].exec(lines[3]);
                if (match) {
                    if (typeof match[1] !== "undefined") status.ControlStateSince = match[1];
                    break;
                }
            }
            status.ControlState = i;
            status.StatusLine = lines[3];
            return status;
        }

        function processLcdText(lines) {
            Q("#status-pane .error").style.display = "none";
            var status = parseLcdText(lines);
            var ModeString = {
                o: "OFF",
                b: "Beer Constant",
                f: "Fridge Const",
                p: "Beer Profile",
                i: "Invalid"
            };
            Object.keys(status).map(function(key, i) {
                var div = Q("#lcd" + key);
                if (div) {
                    if (key == "ControlMode") div.innerHTML = ModeString[status[key]];
                    else if (key == "ControlState") div.innerHTML = (status[key] < STATES.length) ? STATES[status[key]].text : "Unknown State";
                    else if (key.match(/[Temp|Set]$/)) div.innerHTML = status[key] + status.tempUnit;
                    else div.innerHTML = status[key];
                }
            });

        }


`

**Comment by tommueller** [comment]
@vitotai great, thanks. fixed the issue fahrenheit issue and I think showing the temperature unit is a great idea, both is updated in the current version of my repo

**Comment by tommueller** [comment]
I would love for some more people to test and then create a PR in a couple of days! I am also happy if people have suggestions on how to improve the layout!

**Comment by gmoyle** [comment]
I'm interested in testing the new interface, I "thought" I understood how to put it on my brewpiless but no luck. Is there a bin I can just upload ota?

**Comment by tommueller** [comment]
hey @gmoyle . Cool thanks!

There is no bin so far, but uploading is really easy:

- download the files from my fork: https://github.com/tommueller/BrewPiLess/tree/master/htmljs/dist
- go to the filemanager: brewpiless.local:8008/filemanager
- upload the files using the filemanager
  - select index.htm and upload it, do the same for control.htm
  - if you are running the bin that supports gzip [BrewPiLess.v2.3.2.gzip.bin](https://github.com/vitotai/BrewPiLess/blob/master/bins/BrewPiLess.v2.3.2.gzip.bin) you can upload
- reload the brewpiless page, hold the shift button pressed when clicking the button or use ctrl+shift+r to reload page without using the browsers cache

**Comment by gmoyle** [comment]
Done. I do like the new look.
I may be reiterating comments already made, but bear with me. 
 
Some initial comments:

- I won't miss the LCD. Sorry.
- The graph page does not work well on mobile phones (iphone, safari). So that obviously needs some work.
- The first time I clicked on the setup tab link, it actually downloaded the setup.htm file. I went back, refreshed my page, tried again, and it worked properly.
- The Control page takes forever to load. It shows the Communicating with BrewPiLess controller.. message well, infinitely. Until I refresh the page. Then it loads. (chrome on win10)

Some minor GUI improvement suggestions to clean things up.

- The legend header could be adjusted (either font size or language) so that when mousing accros the header doesn't it jump around from one to two rows. It would simply look cleaner. [legendheaderjumping-SWF.zip](https://github.com/vitotai/BrewPiLess/files/1288647/legendheaderjumping-SWF.zip)
- To drive consistency, perhaps the Waiting to Cool, Cooling, Heating text, etc on the STATUS could be color coded to match the colors that used to show under the graph.
- On that point. I miss the heat/cool colors under the graph from the old one.
The tab links are good. I know there was some discussion about them, but they work for me. 
- I would like to see all the pages unified with the same header (as the Graph and Control pages behave)

I will keep hitting it.

Thanks to everyone for all the hard work. I love this new setup, and it was nearly painless to migrate from my raspberry pi setup.

**Comment by tommueller** [comment]
Thanks a lot @gmoyle for the feedback!

Some comments on it:

> The graph page does not work well on mobile phones (iphone, safari). So that obviously needs some work.

I know, this won't be too much work, it's already on my ToDo. Wanted to reach a point of agreement on the general layout first, to avoid duplicate work.

> The first time I clicked on the setup tab link, it actually downloaded the setup.htm file. I went back, refreshed my page, tried again, and it worked properly.

That is ought, let me know if this happens again, otherwise I will assume this is a weird constellation that doesn't need to be investigated (some file-type thing with the backend maybe ...)

> The Control page takes forever to load. It shows the Communicating with BrewPiLess controller.. message well, infinitely. Until I refresh the page. Then it loads.

Did that happen only once? Or everytime you navigate to the page?

> The legend header could be adjusted (either font size or language) so that when mousing accros the header doesn't it jump around from one to two rows. It would simply look cleaner. 

Ah that never happend on my screen, but will fix it!

> To drive consistency, perhaps the Waiting to Cool, Cooling, Heating text, etc on the STATUS could be color coded to match the colors that used to show under the graph.

:+1: 

> On that point. I miss the heat/cool colors under the graph from the old one.

Can you post a screenshot? I'm not sure what you mean.

> The tab links are good. I know there was some discussion about them, but they work for me.
I would like to see all the pages unified with the same header (as the Graph and Control pages behave)

:+1: 

Cheers,
Tom

**Comment by gmoyle** [comment]
> On that point. I miss the heat/cool colors under the graph from the old one.
> Its not critical.. but...

Before...
![2017-09-11_0750](https://user-images.githubusercontent.com/3707172/30281064-4622c130-96c6-11e7-89eb-f9a8b4505fcb.png)

What I am seeing...
![2017-09-11_0755](https://user-images.githubusercontent.com/3707172/30281205-a48f4fe0-96c6-11e7-84e6-f7528025a823.png)

> The Control page takes forever to load. It shows the Communicating with BrewPiLess controller.. message well, infinitely. Until I refresh the page. Then it loads.
> 
> Did that happen only once? Or everytime you navigate to the page?

This only happened 2 or three times. It no longer seems to be an issue. I thought it was strange.

**Comment by tommueller** [comment]
@vitotai do you have an ideao why the heating/cooling bars disappeared? I did not change anything on the graph itself (or so I thougt ;))

**Comment by vitotai** [comment]
@tommueller

That was added recently before you forked the project. You can just update the parts that marked with "chart.js" (line 13 ~ 619).

**Comment by gmoyle** [comment]
I ran into an issue yesterday.  I'm halfway through a brew profile (yes, testing on a live fermentation. Yes, I am crazy) and when I went to look at the profile, the brewpiless seemed frozen. Web interface would not respond. So I cycled the power. Now I cannot get a profile to stick. I can load it, apply it, but when I refresh the page it vanishes. I removed this interface, and reuploaded the latest bin to revert to the tip of the classic interface. Now I am able to load a profile and it sticks.  I don't know if this is the interface or a different issue.

**Comment by tommueller** [comment]
@vitotai thanks I will work that into my code then!

@gmoyle oh that sounds rather bad. kudos for testing on a live fermentation ;)
I think the profile is stil being executed by BPL but my control-page never loads the current profile. I thought this was the default, but apparently it is not?! @vitotai Do you load the current profile into the profile-editor on page-load? Did I accidentally remove that? Or did I missunderstand the issue?

**Comment by allthepies** [comment]
I think I found an issue when running in Beer Profile mode too, the fridge/beer setpoint temperatures weren't being updated on the main page.  This was on my testbed which has no temp probes attached though so not sure if it's a result of that ?

**Comment by vitotai** [comment]
@allthepies

Are  you using Fahrenheit or Celsius? I tested it with a NodeMcu solo, and it worked, except that the control leads to control.html instead of control.htm

**Comment by allthepies** [comment]
Celsius all the way.

**Comment by vitotai** [comment]
@gmoyle
It might have something to do with the filesystem. I am not sure, though.
The beer profile is saved as brewing.json on the file system. The control page load the JSON file and display/edit it. The BPL on ESP8266 opens and reads this file to keep the profile. That says, the front pages have nothing to do with the execution of beer profiles.
 @tommueller 
There is no default beer profile. I just tested the control.htm, and it failed to load the file. I am investigating this issue.

**Comment by tommueller** [comment]
@vitotai aah yes, I also always got that error, that the json file could not be read, but did not know what it meant. Thanks for looking into it.

**Comment by vitotai** [comment]
I found the issue. However, I don't really understand why it works in my original script.
The problem is that the beer profile is encoded by encodeURIComponent() **TWICE**.
The saved file is not correctly formatted, and the beer profile can't be executed, nor can it be reloaded to show.
The solution is remove the encodeURIComponent() in
/script-control.js@line775
`
    BWF.save(BWF.BrewProfile, encodeURIComponent(json), function() {
`
to 
`
    BWF.save(BWF.BrewProfile, json, function() {
`

That should fix this issue.

**Comment by vitotai** [comment]
@allthepies
The display of setting temperature might show after 1 minute delays.
When "Beer Profile" is applied, BrewPi core just set the mode to "Beer Profile". The "Profile Keeper" will check the control status every minute, and the setting temperature is determined when it find that it is in "Beer Profile" mode.

**Comment by allthepies** [comment]
I will fire up the testbed again but I think I left it for some time after selecting Beer Profile.

**Comment by vitotai** [comment]
@allthepies
Please use testcmd.htm, and issue command "l". Copy the result for me. Thanks.

**Comment by tommueller** [comment]
> The legend header could be adjusted (either font size or language) so that when mousing accros the header doesn't it jump around from one to two rows. It would simply look cleaner. legendheaderjumping-SWF.zip

@gmoyle could you maybe reupload that file as normal vid? cannot make my system to play that swf file ...

The other issues mentionend here should be fixed now (except yours @allthepies), still waiting for your output.
There is a ToDo-list on my fork now: https://github.com/tommueller/BrewPiLess/tree/master/htmljs

**Comment by allthepies** [comment]
Loaded up the latest version and all looking good.

However when I load the main page or the control page then I initially get v2.1 displayed with the following alert. 
![image](https://user-images.githubusercontent.com/1494526/30479073-7d68ee4c-9a0b-11e7-92c9-760d9c9bab11.png)

When I confirm the alert then the page goes to v2.3.3
![image](https://user-images.githubusercontent.com/1494526/30479115-abe0dcee-9a0b-11e7-9b73-f4b34eb0eda8.png)


Is this expected while in test mode ?

**Comment by tommueller** [comment]
That is strange, did you manually change the version number on the "server-side" to v2.3.3? I do not get this error with @vitotai bin and cannot see anything (on a short check) in the git history that the version number was bumped.

**Comment by allthepies** [comment]
I took Vito's latest version.  At the bottom of Config.h:

#define BPL_VERSION "2.3.3"

**Comment by tommueller** [comment]
Ah ok, but that is not compiled into the bin he published, that's why I don't get the error.
I will upload a fixed version later.

**Comment by tommueller** [comment]
Allright, new version is up, including mobile support. The graph page now scales dynamically to the screen size. Will try to get to the other pages in the next days.
@allthepies version should now match for you!

**Comment by allthepies** [comment]
The alert has now gone!

However, on the control page there is some text pollution on the top right

![image](https://user-images.githubusercontent.com/1494526/30490549-2f5f7244-9a32-11e7-8f22-8a037592c4d0.png)

**Comment by tommueller** [comment]
yep, did not upgrade the header to the responsive code in control-page. should be fixed now.

**Comment by allthepies** [comment]
That's better.

Love the mobile friendly main page, the control page not so mobile friendly, is that still work in progress ?

And I can't change mode from mobile on the control page, the Apply button greys out when selected but nothing happens and the button remains greyed out.

This is Chrome on Android Marshmallow.  I might hook up the remote debugger and see if any JS errors.

**Comment by tommueller** [comment]
Yes control-page is still WIP. Will do that tomorrow probably.

After pressing the apply button, did you check the graph page? I think it changes the mode. Clicking with touch-devices activates the hover style permantely, but I think it gets upgraded. There should be a notification, when changing was successful though, will add that soon.

**Comment by allthepies** [comment]
Mode didn't change in the main page.

**Comment by vitotai** [comment]
Sorry for the version change. I made some change and plan to make it a version.
- Dedicated TITLE field in config page. I don't like space or special characters in SSID, so a dedicated field for the title might be better.
- updated HTML/JS
- Bug fixed. I found a wrong MIME type in the code which might result in file download instead of rendering.

**Comment by tommueller** [comment]
Sorry I cannot test on mobile devices, I still do not own any form of smartphone or tablet ... maybe you could fire up a debug session @allthepies ?

@vitotai sounds good, was just confused about the version bump

**Comment by allthepies** [comment]
Update:  I fired up the USB debug and..... it worked!  The control page does change the mode as requested.
However the "Apply" button remains greyed out when selected and doesn't return to the normal colour.

**Comment by tommueller** [comment]
Ok great, I will try to find out why the button stays grey and a confirmation when changing mode was succesful.

**Comment by tommueller** [comment]
Ok, control page is responsive as well now. Rest of the pages will follow the next days.

**Comment by vitotai** [comment]
FYI,
I tested on my iPhone and found that mobile safari doesn't handle "Content-Dispostion" well. 
I had to modify the library to make the gzip html work on iPhone.

If you are using iPhone/iPad, you need to use latest code, or the updated BrewPiLess.v2.3.3.bin.

**Comment by gmoyle** [comment]
You guys have been busy the last week. Thanks for making fixes to several of my suggestions. I have 2.3.3 running with the new interface.  Here is a video of the legend title jumping around as you mouse over the graph in mp4. Yes, I know its a small thing. 
[legendtitletoggle.zip](https://github.com/vitotai/BrewPiLess/files/1309255/legendtitletoggle.zip)

**Comment by tommueller** [comment]
Thanks @gmoyle . I see it is the "Waiting for peak" which is too long for the space, I will fix that. Cheers

**Comment by allthepies** [comment]
Has this design concluded now ?

**Comment by allthepies** [comment]
Thanks Tom!

**Comment by tommueller** [comment]
Sorry guys, I am swamped at work these days and cannot find the motivation to keep on coding afterwards. It'll be beginning of November that I'll find time for this. I will let you know when I get back on this to get a status update!

**Comment by vitotai** [comment]
@tommueller 
I made some modification recently. The latest files is put in another branch
https://github.com/vitotai/BrewPiLess/tree/v24-ispindel

Three HTMLs are modified. gdc.htm, setup.htm, and index.htm.
The modification to index.htm is
- add TILT value of iSpindel
- Use short data time format for last update of iSpindel
- add a div to show derived formula.

I will divide the source javascript as what I did before so that it's easier to merge your change.

**Comment by vitotai** [comment]
@tommueller 
I tried to split the HTML/Javascript sources and update them to v2.4. 
Please check this branch.
https://github.com/vitotai/BrewPiLess/tree/v24-ispindel

**Comment by tommueller** [comment]
@vitotai I wanted to continue with the development today, but I am not able to incorporate your changes and get the pages working. I tried merging, rebasing and just copying your htmljs folder over mine, but things are still broken. I did upgrade to 2.4.
Can you maybe merge the projects properly so that I can continue with the other pages? Cheers

**Comment by vitotai** [comment]
I tested and found
- the index.htm works but the "error" div doesn't disappear.  => fixed in script.js
- the controller.htm doesn't work because I changed the init function => fixed in control.html

Sorry about that. I just can't setup the automation environment.

**Comment by tommueller** [comment]
Great, thanks. Now I only have a couple of minor bugs:

- on both pages the current page is not highlighted anymore
- the waiting for update box still does not disappear for me on the graph-page
- is the formula-button supposed to be showing as just an f?
- on the control page the version is not displayed correctly anymore

![auswahl_004](https://user-images.githubusercontent.com/2970691/32671089-5fa87802-c646-11e7-9ef2-ad3a5d8c6a52.png)
![auswahl_005](https://user-images.githubusercontent.com/2970691/32671090-5fd199e4-c646-11e7-9645-82f93fa07bd9.png)

Why can't you install the automation? It's fairly simple ... where does it not work for you?

**Comment by vitotai** [comment]
I had trouble setup sass.

- on both pages the current page is not highlighted anymore
I don't know how you made them highlighted. I even don't notice that.

-  the waiting for update box still does not disappear for me on the graph-page
Should be fixed, this time.

- is the formula-button supposed to be showing as just an f?
Yes. But it should be hidden.

- on the control page the version is not displayed correctly anymore
I didn't notice that. added.

**Comment by tommueller** [comment]
I will doublecheck my instructions, so that you can get it working.

I merged your changes and now everything works, **but** I get a JS error on the control-page

> control.htm:1 Uncaught ReferenceError: displayrssi is not defined
    at Object.V (control.htm:1)
    at Object.process (control.htm:1)
    at EventSource.e.onmessage (control.htm:1)

Don't know why it runs into that case though. That is why the version does not update correctly I guess.

**Comment by vitotai** [comment]
oops. I just couldn't get it right at first place.
Please re-get the script-control.js

I also found:
The highlight of current page is implemented by attaching a "active" class to the anchor, which is in 
header.html.  I didn't change it and have no idea why it's different.

**Comment by tommueller** [comment]
Allright working now!

For some reason the function `getActiveNavItem()` got removed from the shared.js. I readded it, working fine now.

I filed a PR, so that we can maybe have a common ground now before I do further development.

**Comment by tommueller** [comment]
@vitotai the config.htm file is still missing in the folder! could you please add it?

**Comment by tommueller** [comment]
So I finally got round to get this close to the finish line. All pages (except config.htm, don't have that one yet) are now in the same layout. I already reworked the logging-page, the other do have the header and footer, but the content still looks like the old version. I will try to get them done soon.

Happy testing!
Tom

**Comment by tommueller** [comment]
@all for testing please use latest files from my dist folder: https://github.com/tommueller/BrewPiLess/tree/master/htmljs/dist

**Comment by allthepies** [comment]
Thanks Tom, I'll try an do some testing this w/e

**Comment by vitotai** [comment]
@tommueller 
Please update the config.htm
I added a field for HTTP port.

**Comment by tommueller** [comment]
@vitotai 
The config.htm is not yet in the repositoy, you have to add it first.

**Comment by tommueller** [comment]
gravity page has been redone now as well. will do setup and config sometime soon.

**Comment by vitotai** [comment]
@tommueller 

I didn't notice that config.htm is missed. Sorry about that.
script.js is also updated for iSpindel information update. 
I also modified the index.html for adding a new tilt value of iSpindel.

**Comment by tommueller** [comment]
No worries! Thanks for uploading it. Will hopefully finish the 2 pages some time this week, then we can go into testing full on.

**Comment by danielatkins** [comment]
I'll keep an eye on progress, happy to do some testing with my keezer when
you're finished Tom. Unfortunately dont have enough wifi reception for my
wemos to work properly on my fermentation chamber so can't use that just
yet.

**Comment by vitotai** [comment]
Updated script.js for missing temperature unit.

Also, I found that repeating uploading new file on the file system might result in file system corrupt. You might not be able to upload the file. Even the file system isn't corrupt, the performance degrades a lot after a while. Thus, I add some code to format the File System.
The URL to for the file system:
http://brewpiless.local:8008/format-spiffs

If you find that the uploading slows down or incomplete, try to format the file system.

**Comment by vitotai** [comment]
@tommueller 

Some modification for config.htm and gdc.htm(gravity.htm). I didn't think additional "input" and the result will be wrong. Sorry, I am not familiar with github operations.

Some opinions about the gravity.htm:
![gdc](https://user-images.githubusercontent.com/15953213/32818930-3725c564-c97b-11e7-8c87-203f4e7077a5.jpg)

1. the second switch is for "Calibrate iSpindel". The "Tile in Water" field is necessary if"Calibrate iSpindel" is enabled.

2. When there are values of the coefficients, it looks weird and is not clear.

**Comment by tommueller** [comment]
@vitotai just fixed the issues you reported with the gravity page. I don't have a BPL with me right now, so I couldn't test it on a device

**Comment by tommueller** [comment]
Ok, config page is added now as well! As I said, I don't have a BPL with me right now, so untested, please check if everything works (but it's a rather simple page, so I guess that'll be allright!).

Only page missing is setup, which I cannot really do without a BPL. Will do that sometime soonish and then close this issue as it got pretty messy. If there are issues then, please open a new issue.

**Comment by tommueller** [comment]
Sorry guys. I haven't had and won't have time to work on this any further in the next time. I saw that @vitotai copied my version into the new release, I hope that works well. I will close this, because the main discussion is finished anyway.
